### ISSUE TITLE
Crash on redirect when PersistCookieJar is used

### ISSUE THREAD

**Issue by MatrixDev** [cfug/dio#1985]
### Package

cookie_manager

### Version

3.1.0+1

### Operating-System

Android

### Output of `flutter doctor -v`

```shell
flutter doctor -v
[!] Flutter (Channel stable, 3.13.1, on macOS 13.4 22F66 darwin-arm64, locale en-UA)
    • Flutter version 3.13.1 on channel stable at /users/matrixdev/Library/Flutter
    ! Warning: `flutter` on your path resolves to /Users/matrixdev/Library/Flutter/bin/flutter, which is not inside your current Flutter SDK checkout at
      /users/matrixdev/Library/Flutter. Consider adding /users/matrixdev/Library/Flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/matrixdev/Library/Flutter/bin/dart, which is not inside your current Flutter SDK checkout at
      /users/matrixdev/Library/Flutter. Consider adding /users/matrixdev/Library/Flutter/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e1e47221e8 (5 weeks ago), 2023-08-22 21:43:18 -0700
    • Engine revision b20183e040
    • Dart version 3.1.0
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /users/matrixdev/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • ANDROID_HOME = /users/matrixdev/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/matrixdev/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10406996/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.82.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (4 available)
    • A063 (mobile)               • P3127F000299  • android-arm64  • Android 13 (API 33)
    • sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64  • Android 12 (API 32) (emulator)
    • macOS (desktop)             • macos         • darwin-arm64   • macOS 13.4 22F66 darwin-arm64
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 117.0.5938.92

[✓] Network resources
    • All expected network resources are available.
```

### Steps to Reproduce

1. attach CookieManager with PersistCookieJar as an interceptor
2. send a GET request with relative path redirect
3. observe a crash in PersistCookieJar

# Backstack
_checkForErrorResponse (common.dart:55)
_File.open.<anonymous closure> (file_impl.dart:381)
<asynchronous gap>
FileStorage.write (file_storage.dart:87)
<asynchronous gap>
PersistCookieJar._save (persist.dart:183)
<asynchronous gap>
PersistCookieJar.saveFromResponse (persist.dart:117)
<asynchronous gap>
CookieManager._saveCookies (cookie_mgr.dart:126)
<asynchronous gap>

# Why it happens?
### CookieManager._saveCookies
```dart
await cookieJar.saveFromResponse(response.realUri, cookies);
```
### Response.realUri
```dart
// redirects.isNotEmpty = true (1 redirect)
// requestOptions.uri = https://example.com/api/v1/status
// redirects.last.location = /api/v1/status
Uri get realUri => redirects.isNotEmpty ? redirects.last.location : requestOptions.uri;
```
### PersistCookieJar._save
```dart
// host == uri.host == "" (because redirects.last.location is relative)
await storage.write(host, json.encode(_filter(cookies)));
```
### FileStorage.write
```dart
// key == host == ""
final file = File('$_currentDirectory$key');

// CRASH
// FileSystemException: Cannot open file, path = '/data/user/0/com.package.my/files/ie0_ps1/' (OS Error: Is a directory, errno = 21)
await file.writeAsString(value);
```

### Expected Result

Relative redirects should be handled properly

### Actual Result

Request fails with exception:
```
      DioException [unknown]: null
Error: FileSystemException: Cannot open file, path = '/data/user/0/com.package.my/files/ie0_ps1/' (OS Error: Is a directory, errno = 21)
```

**Comment by AlexV525** [comment]
Should be fixed by #1948?

**Comment by MatrixDev** [comment]
@AlexV525 yes, looks like it should. couldn't find a bug for this crash so filled one. dou you know when it will be merged?
